lms/djangoapps/ccx/views.py

```diff
@@ -0,0 +1,526 @@
+    if earliest:
+        override_field_for_ccx(ccx, course, 'start', earliest)
+
+    # Attempt to automatically adjust grading policy
+    changed = False
+    policy = get_override_for_ccx(
+        ccx, course, 'grading_policy', course.grading_policy
+    )
+    policy = deepcopy(policy)
+    grader = policy['GRADER']
+    for section in grader:
+        count = graded.get(section.get('type'), 0)
+        if count < section['min_count']:
+            changed = True
+            section['min_count'] = count
+    if changed:
+        override_field_for_ccx(ccx, course, 'grading_policy', policy)
+
+    return HttpResponse(
+        json.dumps({
+            'schedule': get_ccx_schedule(course, ccx),
+            'grading_policy': json.dumps(policy, indent=4)}),
+        content_type='application/json',
+    )
+
+
+@ensure_csrf_cookie
+@cache_control(no_cache=True, no_store=True, must_revalidate=True)
+@coach_dashboard
+def set_grading_policy(request, course):
+    """
+    Set grading policy for the CCX.
+    """
+    ccx = get_ccx_for_coach(course, request.user)
+    override_field_for_ccx(
+        ccx, course, 'grading_policy', json.loads(request.POST['policy']))
+
+    url = reverse('ccx_coach_dashboard', kwargs={'course_id': course.id})
+    return redirect(url)
+
+
+def validate_date(year, month, day, hour, minute):
+    # avoid corrupting db if bad dates come in
+    valid = True
+    if year < 0:
+        valid = False
+    if month < 1 or month > 12:
+        valid = False
+    if day < 1 or day > 31:
+        valid = False
+    if hour < 0 or hour > 23:
+        valid = False
+    if minute < 0 or minute > 59:
+        valid = False
+    return valid
+
+
+def parse_date(datestring):
+    """
+    Generate a UTC datetime.datetime object from a string of the form
+    'YYYY-MM-DD HH:MM'.  If string is empty or `None`, returns `None`.
+    """
+    if datestring:
+        date, time = datestring.split(' ')
+        year, month, day = map(int, date.split('-'))
+        hour, minute = map(int, time.split(':'))
+        if validate_date(year, month, day, hour, minute):
+            return datetime.datetime(
+                year, month, day, hour, minute, tzinfo=pytz.UTC)
+
+    return None
+
+
+def get_ccx_for_coach(course, coach):
+    """
+    Looks to see if user is coach of a CCX for this course.  Returns the CCX or
+    None.
+    """
+    try:
+        return CustomCourseForEdX.objects.get(
+            course_id=course.id,
+            coach=coach)
+    except CustomCourseForEdX.DoesNotExist:
+        return None
+
+
+def get_ccx_schedule(course, ccx):
+    """
+    Generate a JSON serializable CCX schedule.
+    """
+    def visit(node, depth=1):
+        """
+        Recursive generator function which yields CCX schedule nodes.
+        """
+        for child in node.get_children():
+            start = get_override_for_ccx(ccx, child, 'start', None)
+            if start:
+                start = str(start)[:-9]
```

Comment:
Not sure why GitHub didn't do the "outdated diff" thing here on this comment, but the current PR has a longer docstring here:

```
Recursive generator function which yields CCX schedule nodes.
We convert dates to string to get them ready for use by the js date
widgets, which use text inputs.
```

See https://github.com/edx/edx-platform/pull/6636/files#diff-faebd8eb1d591f3ad3147fe858e4e849R286